workflow-bot-app/src/main/java/com/symphony/bdk/workflow/engine/executor/request/RequestExecutor.java

```diff
@@ -1,49 +1,62 @@
 import com.symphony.bdk.workflow.swadl.v1.activity.request.ExecuteRequest;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
-import com.fasterxml.jackson.databind.ObjectMapper;
 import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
 
-import java.util.Collections;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 @Slf4j
+@Component
 public class RequestExecutor implements ActivityExecutor<ExecuteRequest> {
 
   private static final String OUTPUT_STATUS_KEY = "status";
   private static final String OUTPUT_BODY_KEY = "body";
 
+  private final HttpClient httpClient;
+
+  public RequestExecutor(HttpClient httpClient) {
+    this.httpClient = httpClient;
+  }
+
   @Override
   public void execute(ActivityExecutorContext<ExecuteRequest> execution) throws JsonProcessingException {
     ExecuteRequest activity = execution.getActivity();
-    Object data;
-    int statusCode;
 
     try {
-      ApiResponse<Object> apiResponse = execution.bdk()
-          .apiClient(activity.getUrl())
-          .invokeAPI("", activity.getMethod(), Collections.emptyList(), activity.getBody(), activity.getHeaders(),
-              Collections.emptyMap(), Collections.emptyMap(), "application/json", "application/json", null,
-              new TypeReference<>() {});
-
-      data = apiResponse.getData();
-      statusCode = apiResponse.getStatusCode();
-    } catch (ApiException apiException) {
-      statusCode = apiException.getCode();
-
-      // we need to set the responseBody as a map instead of a string to allow processing it in subsequent activities
-      data = new ObjectMapper().readValue(apiException.getResponseBody(), Map.class);
-      log.debug("This error happens when the request fails.", apiException);
+      Response response =
+          this.httpClient.execute(activity.getMethod(), activity.getUrl(), activity.getBody(),
+              headersToString(activity.getHeaders()));
+
+      execution.setOutputVariable(OUTPUT_STATUS_KEY, response.getCode());
+      execution.setOutputVariable(OUTPUT_BODY_KEY, response.getContent());
+    } catch (IOException ioException) {
+      log.debug("This error happens when the request fails with IOException.", ioException);
+      throw new RuntimeException(ioException.getMessage());
     }
+  }
 
-    execution.setOutputVariable(OUTPUT_STATUS_KEY, statusCode);
-    execution.setOutputVariable(OUTPUT_BODY_KEY, data);
+  private Map<String, String> headersToString(Map<String, Object> headers) {
+    Map<String, String> result = new HashMap<>();
+    for (Map.Entry<String, Object> entry : headers.entrySet()) {
+
+      // We consider that opening/closing brackets are not allowed in headers values.
+      // Otherwise, the list items should be added in the string one by one using map function
+      String entryListAsString = List.of(entry.getValue()).toString();
```

Comment:
is it covered by tests?

can't we support a map of arrays if we want to support repeated headers? (to avoid parsing the strings)

Reply:
In ExecuteRequestIntegrationTest.java: 

**SWADL with a header having a list of items:** https://github.com/finos/symphony-wdk/pull/58/files#diff-57f7d5d71b9c4de5368b1d6e2e033fda26423133a9b0ebd5f20694ed9e21c0eeR15
**Expected map to be used:**  https://github.com/finos/symphony-wdk/pull/58/files#diff-26eea4bd568d90ac0dc9e316804f84a2f6f7d01bea0f8fb2d1b017e5b0dc7ce4R43

Reply:
If we consider headers as a map of arrays then a header that only has 1 value is considered as a list of 1 item, or am I wrong? 
In this case we will have to set headers in SWADL as a list (of 1 item when not repeated).

Reply:
Good, we have tests for it and the YAML looks good, but then why do we have to parse/remove []?
Can't we just take all the values and join them if they are array items?
it looks like what we do relies on the .toString() of a list (that might not be a list), it seems fragile

Reply:
`it looks like what we do relies on the .toString() of a list (that might not be a list), it seems fragile`:
I enforce it to be a list. _.toString()_ is called on _List.of(header).toString()_ ==> So in case the header is not a list we enforce it to be so before calling toString()

`Can't we just take all the values and join them if they are array items?`
By using "instance of"? so we join them only if the object is an instance of list? I tried to avoid using that.

The issue here is that String.joins() requires a CharSequence parameter and we cannot cast List -> CharSequence. 
I agree that this seems fragile,

Reply:
```
if (entry.getValue() instanceof List) {
      StringBuilder stringBuilder = new StringBuilder();
      ((List) entry.getValue()).forEach(item -> stringBuilder.append(item).append(","));

      result.put(entry.getKey(), String.join(",", stringBuilder.deleteCharAt(stringBuilder.length() - 1)));
} else {
      result.put(entry.getKey(), (String) entry.getValue());
}
```

Reply:
WDYT?  (I was avoiding to use instance of but I don't see a better way)

```
  private Map<String, String> headersToString(Map<String, Object> headers) {
    Map<String, String> result = new HashMap<>();
    for (Map.Entry<String, Object> entry : headers.entrySet()) {
      StringBuilder stringBuilder = new StringBuilder();
      toList(entry.getValue()).forEach(item -> stringBuilder.append(item).append(","));
      result.put(entry.getKey(), stringBuilder.deleteCharAt(stringBuilder.length() - 1).toString());
    }
    return result;
  }

  @SuppressWarnings("unchecked")
  private List<String> toList(Object object) {
    if (object instanceof String) {
      return List.of((String) object);
    } else {
      return new ArrayList<>(((List<String>) object));
    }
  }
```